key-share/src/serde_fix.rs

```diff
@@ -0,0 +1,42 @@
+//! Fixes serde serialization of the key share
+//!
+//! This module contains structs that help us implementing `serde::{Serialize, Deserialize}`
+//! for [`crate::DirtyCoreKeyShare`].
+//!
+//! Context: we used to have key share struct flatten, with all DirtyKeyInfo fields being
+//! right in the key share. However, at some point, we've decided to move all public common
+//! fields into a separate structure. In order to keep serialization format compatible,
+//! we used `#[serde(flatten)]` attribute, to make on-wire data appear like all the fields
+//! are still in the same struct. It turned out, that `#[serde(flatten)]` is buggy, and,
+//! specifically, it does not preserve `is_human_readable` flag, which broke (de)serialization
+//! code and compatibility with old key shares.
+//!
+//! See the issue for more details on the `serde` problem: <https://github.com/serde-rs/serde/issues/2704>
+//!
+//! Until the issue in `serde` crate is addressed, we have to use a workaround in this module.
+//! We basically reimplement `flatten` attribute manually at the cost of extra allocations.
+
+use generic_ec::{serde::CurveName, Curve, NonZero, Point, SecretScalar};
+use serde::{Deserialize, Serialize};
+use serde_with::As;
+
+#[derive(Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct CoreKeyShare<E: Curve> {
+    pub curve: CurveName<E>,
+    pub i: u16,
+    #[serde(with = "As::<generic_ec::serde::Compact>")]
+    pub shared_public_key: NonZero<Point<E>>,
+    #[serde(with = "As::<Vec<generic_ec::serde::Compact>>")]
+    pub public_shares: Vec<NonZero<Point<E>>>,
+    pub vss_setup: Option<crate::VssSetup<E>>,
+    #[cfg(feature = "hd-wallets")]
+    #[cfg_attr(
+        feature = "serde",
+        serde(default),
+        serde(with = "As::<Option<crate::utils::HexOrBin>>")
+    )]
+    pub chain_code: Option<slip_10::ChainCode>,
+    #[serde(with = "As::<generic_ec::serde::Compact>")]
+    pub x: NonZero<SecretScalar<E>>,
+}
```

Comment:
Can this struct instead have all it fields be references? This way there will be no allocations. I'm 50% sure I did something similar for serde once.

Reply:
There would need to be a second struct with non-reference fields for deserialization though

Reply:
Yeah we can do this, but I didn't want to duplicate the struct and have serialization and deserialization in different places

Reply:
On one hand, I don't see problem with a little duplicating for a great cause. On the other, I think cloning here is cheap and acceptable. I leave the choice to you

Reply:
I just tried to duplicate the structs, and what I don't like in this approach is that it's difficult to track all the serde attributes we have. They need to be the same, so the serialization is compatible with deserialization, but when they're duplicated, it's difficult to check if there are any differences.

Reply:
I'll see if the macro helps

Reply:
Yep it looks good (I had to do a small change in generic-ec tho)